### PR TITLE
Respect ROCR_VISIBLE_DEVICES on AMD GPU device discovery

### LINKED ISSUE #140318: Visible devices are not respected on AMD systems
### 🐛 Describe the bug

When running distributed (or multi-process) jobs on AMD GPU systems with GPU isolation (i.e., the `*_VISIBLE_DEVICES` environment variable), PyTorch still tries to set the seed of all the devices it finds. This causes the following exception to be raised:

```
File "/path/to/torch/random.py", line 46, in manual_seed                                                                                               
  torch.cuda.manual_seed_all(seed)                                                                                                                                                                      
File "/path/to/torch/cuda/random.py", line 131, in manual_seed_all                                                                                    
  _lazy_call(cb, seed_all=True)                                                                                                                                                                        
File "/path/to/torch/cuda/__init__.py", line 249, in _lazy_call                                                                                       
  callable()                                                                                                                                                                                            
File "/path/to/torch/cuda/random.py", line 128, in cb                                                                                                  
  default_generator = torch.cuda.default_generators[i]                                                                                                                                                 
                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^                                                                                                                                                 
IndexError: tuple index out of range
```

The underlying issue is that the visible devices are not properly detected due to reading the wrong environment variable name: `HIP_VISIBLE_DEVICES` should be renamed to `ROCR_VISIBLE_DEVICES` (see https://rocm.docs.amd.com/en/latest/conceptual/gpu-isolation.html#rocr-visible-devices ). On many batch schedulers and clusters, the ROCr environment variable is the one that is set.

This was likely introduced in #119182 (cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd)

I will push a pull request to fix this later today.

### Versions

PyTorch version: 2.5.1+rocm6.2

### PR DESCRIPTION
**Pull request by tbennun**
Reland of #140320 after failing test on trunk. Fixes potential environment clobbering in test, makes ROCr+HIP devices (if specified together) more robust to index errors.

Fixes #140318

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd

### CODE REVIEW

**Comment by tbennun** [comment]
FYI (previous approvers) @jataylo @jithunnair-amd @eqy @jeffdaily

File: torch/cuda/__init__.py
Comment (jataylo) [inline torch/cuda/__init__.py:659]: If I understand this correctly, I don't believe this assumption to be true.

For example we can set ROCR_VISIBLE_DEVICES="6,7" and HIP_VISIBLE_DEVICES="1,0"

This informs runtime to expose device 6,7 and the hip application has freedom to set these as they wish. 

And as far as I'm aware if we set 

ROCR_VISIBLE_DEVICES="4" HIP_VISIBLE_DEVICES="4" this would cause issues and HIP_VISIBLE_DEVICES wouldn't see a device 4

**Review by jataylo — CHANGES_REQUESTED** [review]

File: torch/cuda/__init__.py
Comment (jataylo) [reply torch/cuda/__init__.py:659]: Also HIP_VISIBLE_DEVICES can accept UUID input, so if a user uses ROCR_VISIBLE_DEVICES to expose certain GPUs and then uses HIP_VISIBLE_DEVICES with UUID input then this would also not be a subset.

**Review by jataylo — CHANGES_REQUESTED** [review]

File: torch/cuda/__init__.py
Comment (jataylo) [reply torch/cuda/__init__.py:659]: Were there any issues in the original way I proposed to handle this? https://github.com/pytorch/pytorch/pull/140320/commits/9afc7f29a34080843042607d237780e3570c4511

Comment (jataylo) [reply torch/cuda/__init__.py:659]: We don't really need to parse ROCR_VISIBLE_DEVICES inside PyTorch this is runtime level control of device selection, but what we need to ensure is that the length of ROCR_VISIBLE_DEVICES is respected to avoid the issue in https://github.com/pytorch/pytorch/issues/140318

Comment (ausellis0) [reply torch/cuda/__init__.py:659]: Hi @jataylo, 
Just wanted to leave a small note, documentation here says `HIP_VISIBLE_DEVICES` only accepts indices: https://rocm.docs.amd.com/en/latest/conceptual/gpu-isolation.html#hip-visible-devices
Maybe that documentation is stale? Thanks!

Comment (jataylo) [reply torch/cuda/__init__.py:659]: Since 6.2 HIP_VISIBLE_DEVICES supports UUID input, we should reach out to update docs.

Comment (jeffdaily) [reply torch/cuda/__init__.py]: I pushed an update to this logic.  The current logic was wrong; Jack was correct that we don't need to parse or index into ROCR_VISIBLE_DEVICES at all.

Comment (tbennun) [reply torch/cuda/__init__.py]: Thanks! I just want to confirm the behavior when both environment variables are set before correcting the PR. The problem lies in the fact that the function should return a list of indices/strings, so returning an arbitrary list (albeit of the correct length) would work for getting the number of devices, but may fail in cases where the function's output would actually be used.

File: torch/cuda/__init__.py
Comment (tbennun) [inline torch/cuda/__init__.py:656]:
```suggestion
            rocr_count = len(rocr_devices.split(","))
```

File: torch/cuda/__init__.py
Comment (tbennun) [inline torch/cuda/__init__.py:658]: could HIP_VISIBLE_DEVICES be set to `0,GPU-uuid,2`? If so, this check should be on `any(dev for dev in hip_devices.split(","))`

File: torch/cuda/__init__.py
Comment (tbennun) [inline torch/cuda/__init__.py:660]:
```suggestion
                        "Cannot specify both ROCR_VISIBLE_DEVICES "
```
nit

File: torch/cuda/__init__.py
Comment (tbennun) [inline torch/cuda/__init__.py:670]: What if I set only `ROCR_VISIBLE_DEVICES` to `3`? Shouldn't the result be the fourth GPU rather than the first?

Comment (jataylo) [reply torch/cuda/__init__.py:670]: @tbennun  This is returning a list of devices equivalent to the amount of devices supplied via ROCR_VISIBLE_DEVICES, the index supplied by ROCR_VISIBLE_DEVICES doesn't matter here as its the runtime that does the isolation not pytorch

E.g. for ROCR_VISIBLE_DEVICES="4,5,6" this logic will return "0,1,2" for indexing devices in the application layer, this will result in the exact same gpu mapping as specified with ROCR_VISIBLE_DEVICES.

File: torch/cuda/__init__.py
Comment (jataylo) [reply torch/cuda/__init__.py:658]: @jeffdaily Although not a common use case it could be possible for the user to use  ROCR_VISIBLE_DEVICES in combination with HIP_VISIBLE_DEVICES with UUIDs

e.g. may use ROCR to expose only GPUs "0,1" in some environment but the user in that environment may want to use UUIDs in HIP_VISIBLE_DEVICES to ensure they run on the requested gpu.

File: torch/cuda/__init__.py
Comment (tbennun) [reply torch/cuda/__init__.py:670]: I see what you mean, that clears it up. Thanks!

**Comment by tbennun** [comment]
@jeffdaily @jataylo I think this is ready for testing before merging, apart from @jataylo 's comment on mixing UUIDs and indices in `HIP_VISIBLE_DEVICES`.

File: torch/cuda/__init__.py
Comment (jeffdaily) [reply torch/cuda/__init__.py]: I was wrong about ROCR env var only taking ints.  It can take UUIDs also.  So ROCR and HIP env vars are similar today, but ROCR env var activates first lower-level than HIP env var so it will reduce the number of devices that HIP can select from.  I think the best we can do now is go back to the simple check that HIP env var isn't asking for more devices than are visible.

**Review by jataylo — APPROVED** [review]
Looks good to me, we're good to go. Lets run CI.

(Slight unknown to me is how ROCR_VISIBLE_DEVICES will operate inside the os.env inside the UT, but if we see some issues there lets just remove the test)

**Comment by jataylo** [comment]
@jeffdaily can we get your approval then lets reland this and open a cherry pick for release 2.6

**Comment by jataylo** [comment]
@atalman @huydhn Hi both, could you help us re-land this PR, we would like to get this bug fix into 2.6 cherry picks also.

**Review by huydhn — APPROVED** [review]

**Review by jeffdaily — APPROVED** [review]

**Comment by jataylo** [comment]
Taking a look at the release

**Comment by jataylo** [comment]
{"CUDA_VISIBLE_DEVICES": "0", "HIP_VISIBLE_DEVICES": None},   **- Returns 8** 
{"CUDA_VISIBLE_DEVICES": None, "HIP_VISIBLE_DEVICES": "0"},    - Returns 1 
{"CUDA_VISIBLE_DEVICES": "0,1,2,3", "HIP_VISIBLE_DEVICES": "0"}, - Returns 1 
{"ROCR_VISIBLE_DEVICES": "1,2,3", "HIP_VISIBLE_DEVICES": "0"}, - Returns 1 
{"ROCR_VISIBLE_DEVICES": "0", "HIP_VISIBLE_DEVICES": None}, - Returns 1 

This issue is we broke CUDA_VISIBLE_DEVICES functionality. Filing a fix and will attempt to remerge.

**Pull request closed without merging**
